### PR TITLE
pjmedia: apply the negotiated max bitrate to the OpenH264 encoder at open

### PR DESCRIPTION
**Pull request by sauwming**
Split out from #5107, which bundles two unrelated changes. This is the bitrate half only; the packetization mode change in that PR is left alone and still needs a separate decision.

### Problem

`oh264_codec_open()` never sets `iMaxBitrate`, so the encoder keeps OpenH264's default of `UNSPECIFIED_BIT_RATE` and runs with no rate control ceiling at all. The negotiated `max_bps` only ever reaches the encoder through `oh264_codec_modify()`, which sets `ENCODER_OPTION_MAX_BITRATE` — and many calls never trigger a bitrate modify, so they run their entire duration unconstrained by what was agreed in SDP.

This matters more since #5121, which derives the video send pacing budget from `max_bps`. An encoder free to burst past that budget is exactly what accumulates pacing debt.

### Changes

- Seed `iMaxBitrate`, and the matching per-layer `iMaxSpatialBitrate`, from `enc_fmt.det.vid.max_bps` at codec open.
- OpenH264 rejects a parameter set whose ceiling is below the target. Where the negotiated maximum is lower than the average (`b=TIAS`/`b=AS` or `max-br` can clamp `max_bps` down without touching `avg_bps`), lower the target to the maximum rather than raise the ceiling above what the remote agreed to receive. The codec param is updated to match so `oh264_codec_get_param()` keeps reporting what is actually in effect.
- Raise `DEFAULT_MAX_BITRATE` to 320000, 25% above `DEFAULT_AVG_BITRATE`.

### Why the default had to change

`DEFAULT_AVG_BITRATE` and `DEFAULT_MAX_BITRATE` were both 256000, and `max_bps` is only ever clamped *downward* from the codec default (`vid_stream_info.c:181`, `vid_codec_util.c:722`). So on the default path the ceiling would have landed exactly on the target — zero burst headroom, with `bEnableFrameSkip = 1` already enabled. Keyframes and scene changes are where that shows. `vid_toolbox.m` already defaults its maximum (512000) above its average (384000), so this follows existing practice in the tree.

Note this raises the `b=TIAS` we advertise for OpenH264 from 256000 to 320000, since that line carries the maximum of the codecs' `max_bps` (`endpoint.c:912`).

### Testing

`pjmedia` builds clean with `-Wall`. `openh264.cpp` is not compiled by the default configure here, so it was additionally checked against the OpenH264 2.3.0 headers with `g++ -fsyntax-only -Wall -DPJMEDIA_HAS_OPENH264_CODEC=1` — no errors, no warnings. Not runtime-tested against a live OpenH264 call.

Co-Authored-By: Claude Opus 5 <noreply@anthropic.com>

### CODE REVIEW

**Comment by sauwming** [comment]
## Review received

> **1. `max_bps == 0` / genuinely small max:** `iMaxBitrate = max_bps` and `iMaxSpatialBitrate = max_bps` are written unconditionally (only the conflict branch is guarded by `> 0`, `openh264.cpp:528`). If `max_bps` can be 0, OpenH264 may interpret this as a hard zero ceiling rather than "unspecified". The author only `-fsyntax-only` checked against 2.3.0 headers with no runtime test — this edge case is exactly what runtime testing would catch. Recommend confirming 0-handling or guarding `iMaxBitrate`/`iMaxSpatialBitrate` to the OpenH264 "unspecified" value when `max_bps <= 0`.
>
> **2. Wire-visible `b=TIAS` bump (256k→320k advertised):** acknowledged in the description, reasonable, but it does raise what remote endpoints see for OpenH264 calls.

## Response

Both points addressed. I did the runtime testing rather than reasoning from headers, and it turned up a real bug — thanks for pushing on this.

### 1. `max_bps == 0` and genuinely small `max_bps`

**`max_bps == 0` is safe, and deliberately so.** `UNSPECIFIED_BIT_RATE` is literally `0` (`codec_app_def.h:60`), so writing `max_bps` through unconditionally *is* the "unspecified" value — there is no separate sentinel to guard toward. In `encoder_ext.cpp:88-93` a zero `iMaxSpatialBitrate` is expanded to the level ceiling rather than treated as a hard cap. Confirmed at runtime: `max=0` initializes fine and the effective layer ceiling comes back as **288000000** (the LEVEL_5_2 limit), i.e. no meaningful ceiling — identical to today's behavior. I've used the `UNSPECIFIED_BIT_RATE` symbol instead of a bare `0` and documented this, since the intent clearly wasn't obvious from the code.

**The "genuinely small max" half was a real bug.** `ParamValidationExt()` rejects a target below one bit per frame (`encoder_ext.cpp:74-77`), so lowering the target to a tiny negotiated maximum made `InitializeExt()` fail outright and the stream get no video at all. Fixed in a9425a2: such a ceiling is now ignored with a warning, so a bogus `b=TIAS` degrades pacing instead of killing the video stream.

Runtime results, mirroring `oh264_codec_open()`'s exact parameter setup against libopenh264 2.3.0:

```
--- with this PR ---
new default avg=256k max=320k     target=256000  max=320000  -> OK   (layer_max=320000)
old default avg=256k max=256k     target=256000  max=256000  -> OK   (layer_max=256000)
max_bps unset (0)                 target=256000  max=0       -> OK   (layer_max=288000000)
remote b=TIAS lower than avg      target=128000  max=128000  -> OK   (layer_max=128000)
pathological max_bps=10 (< fps)   target=256000  max=0       -> OK   (ceiling ignored)

--- before the a9425a2 guard ---
pathological max_bps=10 (< fps)   target=10      max=10      -> FAIL (ENC_RETURN_UNSUPPORTED_PARA)

--- max < target passed through raw, no clamp at all ---
                                  target=256000  max=128000  -> FAIL (ENC_RETURN_UNSUPPORTED_PARA)
```

Two things worth noting from that last row: handling the max-below-target conflict is *mandatory*, not cosmetic — without it codec open fails. And #5107's clamp direction (raise the ceiling to the target) initializes fine but transmits above what the remote negotiated, which is why this PR lowers the target instead.

Separately, the runtime run also confirms the `DEFAULT_MAX_BITRATE` bump was necessary rather than cosmetic: with `max == avg` OpenH264 itself logs *"Setting MaxSpatialBitrate the same at SpatialBitrate will make the actual bit rate lower than SpatialBitrate"* (`encoder_ext.cpp:112-115`).

### 2. `b=TIAS` 256k → 320k

Confirmed, and it stays. Worth adding that this isn't a new class of behavior: `vid_toolbox.m` already defaults `max_bps` to 512000, and `endpoint.c:912` takes the maximum across codecs' `max_bps` — so any build with VideoToolbox already advertises 512000 today. OpenH264 was the outlier in advertising a ceiling equal to its own target, which is the thing that made the fix unsafe without the bump.

## Modification made

Commit a9425a2, the only change since the review:

```diff
     eprm.iMaxBitrate                    = param->enc_fmt.det.vid.max_bps;
-    if (eprm.iMaxBitrate > 0 && eprm.iMaxBitrate < eprm.iTargetBitrate) {
-        eprm.iTargetBitrate             = eprm.iMaxBitrate;
-        param->enc_fmt.det.vid.avg_bps  = eprm.iMaxBitrate;
+    if (eprm.iMaxBitrate != UNSPECIFIED_BIT_RATE &&
+        eprm.iMaxBitrate < eprm.iTargetBitrate)
+    {
+        if ((float)eprm.iMaxBitrate >= eprm.fMaxFrameRate) {
+            eprm.iTargetBitrate         = eprm.iMaxBitrate;
+            param->enc_fmt.det.vid.avg_bps = eprm.iMaxBitrate;
+        } else {
+            /* OpenH264 rejects a target below one bit per frame, which would
+             * fail codec open. Drop the ceiling instead of the whole stream.
+             */
+            PJ_LOG(3,(THIS_FILE, "Ignoring unusable max bitrate %d bps "
+                                 "(below %d fps)", eprm.iMaxBitrate,
+                                 (int)eprm.fMaxFrameRate));
+            eprm.iMaxBitrate            = UNSPECIFIED_BIT_RATE;
+        }
     }
```

plus a comment line recording that an unset `max_bps` is itself `UNSPECIFIED_BIT_RATE`, so it passes through as "no ceiling".

## Testing caveat

The probe above exercises the encoder parameter path only — `InitializeExt()` plus `GetOption()` readback against the real library. It is not a full call test, so the pacing interaction with #5121 is still reasoned rather than measured. `pjmedia` builds clean, and `openh264.cpp` compiles warning-free with `-Wall` against the 2.3.0 headers (it isn't built by the default configure here).

File: pjmedia/src/pjmedia-codec/openh264.cpp [hunk 1]
```diff
@@ -511,6 +515,33 @@ static pj_status_t oh264_codec_open(pjmedia_vid_codec *codec,
...
+            eprm.iTargetBitrate         = eprm.iMaxBitrate;
+            param->enc_fmt.det.vid.avg_bps = eprm.iMaxBitrate;
+        } else {
+            /* OpenH264 rejects a target below one bit per frame, which would
+             * fail codec open. Drop the ceiling instead of the whole stream.
+             */
+            PJ_LOG(3,(THIS_FILE, "Ignoring unusable max bitrate %d bps "
+                                 "(below %d fps)", eprm.iMaxBitrate,
+                                 (int)eprm.fMaxFrameRate));
+            eprm.iMaxBitrate            = UNSPECIFIED_BIT_RATE;
```
[the remark this answers is not in the document]
Comment (sauwming) [reply pjmedia/src/pjmedia-codec/openh264.cpp]: Good catch — fixed in b96cc12. Both halves of this were right: dropping the ceiling did leave the encoder unconstrained, and it left `get_param()` reporting a ceiling that was no longer in effect.

I went with the "preserve an enforceable ceiling" option rather than failing codec open. Failing would be a robustness regression relative to master, where a nonsensical `b=TIAS` is simply ignored and the call still gets video — refusing to open the codec over a bad bandwidth line trades a cosmetic problem for a dead video stream.

So the ceiling is now raised to the lowest rate the encoder accepts and applied, instead of being discarded:

```c
if ((float)eprm.iMaxBitrate < eprm.fMaxFrameRate) {
    PJ_LOG(3,(THIS_FILE, "Max bitrate %d bps is below the encoder "
                         "minimum, raising it to %d bps",
                         eprm.iMaxBitrate, (int)eprm.fMaxFrameRate + 1));
    eprm.iMaxBitrate            = (int)eprm.fMaxFrameRate + 1;
}
eprm.iTargetBitrate             = eprm.iMaxBitrate;
param->enc_fmt.det.vid.avg_bps  = eprm.iMaxBitrate;
param->enc_fmt.det.vid.max_bps  = eprm.iMaxBitrate;
```

The stream stays constrained as close to the negotiated limit as the encoder can express, and `max_bps` is written back so `oh264_codec_get_param()` reports what is actually in effect — which was the second half of your point.

The `+ 1` is deliberate: `ParamValidationExt()` compares the target against `fFrameRate` as a float, so truncating a fractional rate like 29.97 to 29 lands just under the limit and fails.

Verified at runtime against libopenh264 2.3.0, driving `InitializeExt()` with `oh264_codec_open()`'s exact parameter setup:

```
pathological max_bps=10, fps=15      target=16     max=16     -> OK  (layer_max=16)
pathological max_bps=10, fps=29.97   target=30     max=30     -> OK  (layer_max=30)
max_bps=29 just under fps=29.97      target=30     max=30     -> OK  (layer_max=30)
max_bps unset (0)                    target=256000 max=0      -> OK  (unspecified, level ceiling)
remote b=TIAS lower than avg         target=128000 max=128000 -> OK  (layer_max=128000)
```

No path now leaves the encoder unconstrained except `max_bps == 0`, which is OpenH264's own `UNSPECIFIED_BIT_RATE` and means "no ceiling was negotiated" — the same reading `vid_stream.c:2288-2290` already applies to it.

**Review by nanangizz — APPROVED** [review]
